packages/jmix-front-generator/src/building-blocks/stages/template-model/pieces/entity-cards/cards.ts

```diff
@@ -0,0 +1,36 @@
+import { EntityAttribute, ProjectModel } from "../../../../../common/model/cuba-model";
+import { getDisplayedAttributes, ScreenType } from "../../../../../generators/react-typescript/common/entity";
+import { EntityWithPath } from "../entity";
+import { getTopAttributesFromQuery } from "../getTopAttributesFromQuery";
+
+export type AttributesTemplateModel = {
+    attributes: EntityAttribute[];
+  }
+  
+export type QueryAnswers = {
+    entity: EntityWithPath;
+    query: string;
+}
+
+/**
+ * To be used in GraphQL-based templates that are displaying entities
+ *
+ * @param answers
+ * @param projectModel
+ */
+ export const deriveAttributesFromQuery = (
+    answers: QueryAnswers, projectModel: ProjectModel
+  ): AttributesTemplateModel => {
+    const attributes = getTopAttributesFromQuery(answers.query);
+  
+    const viewProps = attributes.map(name => ({name}));
+  
+    return {
+      attributes: getDisplayedAttributes(
+        viewProps,
+        answers.entity,
+        projectModel,
+        ScreenType.BROWSER
+      )
+    };
+  }
```

Comment:
No need to duplicate this code, stuff that is located inside `building-blocks` can be freely reused between generators, so we can just use the code from `pieces/entity-browser/browser.ts`. Cards template is kind of entity browser so I think the naming is OK, but we can also change `entity-browser` to something more generic.